### PR TITLE
Upgrade spl-token to v0.3.8

### PR DESCRIPTION
**Pull request by odcheung**
https://app.asana.com/0/1202413563699470/1204278477629625/f

- bump common-sdk to 0.3.0 (spl-token upgrade)
- update js target to ES2020 to use bigInt literals
- replace all deriveATA to getAssociatedTokenAddressSync
- replace all spl-token instruction builders previously under the Token namespace
- Replace all u64 with BN
- Update parser to use the EntityParser stack (TokenAccount & Mint parsers) from `common-sdk`. Update parser method (breaking) to include `address`
- Replace old Token types `AccountInfo` & `MintInfo` with `Account` & `Mint`
- Remove duplicated account-request util methods. Use the ones in `common-sdk`
- Remove duplicated util methods `createATAIx`, `getAssociatedTokenAddressSync`, `createWSOLAccountInstructions`

## Verification
- Verified by running the unit-test suite
- Will upgrade and verify Orca2 w/ this release prior to releasing

### CODE REVIEW

File: sdk/tests/sdk/whirlpools/position-impl#collectFees.test.ts [hunk 1]
```diff
@@ -178,13 +179,13 @@ describe("PositionImpl#collectFees()", () => {
 
       assert.notEqual(positionDataAfter, null);
 
-      const accountAPubkey = await deriveATA(otherWallet.publicKey, poolInitInfo.tokenMintA);
-      const accountA = await testCtx.whirlpoolCtx.fetcher.getTokenInfo(accountAPubkey, true);
-      assert.ok(accountA && accountA.amount.eq(quote.feeOwedA));
+      const accountAPubkey = getAssociatedTokenAddressSync(poolInitInfo.tokenMintA, otherWallet.publicKey);
+      const accountA: Account = await testCtx.whirlpoolCtx.fetcher.getTokenInfo(accountAPubkey, true);
```
Comment (yugure-orca) [inline sdk/tests/sdk/whirlpools/position-impl#collectFees.test.ts:183]: super nit: is ``: Account``  required?  (accountBPubkey is defined explicit data type)

File: sdk/tests/sdk/whirlpools/whirlpool-impl#closePosition.test.ts [hunk 2]
```diff
@@ -508,9 +509,10 @@ describe("WhirlpoolImpl#closePosition()", () => {
 
       const position = await testCtx.whirlpoolClient.getPosition(positionData.publicKey, true);
 
-      const walletPositionTokenAccount = await deriveATA(
+      const walletPositionTokenAccount = getAssociatedTokenAddressSync(
+
```
Comment (yugure-orca) [inline sdk/tests/sdk/whirlpools/whirlpool-impl#closePosition.test.ts:513]: nit: blank line ?

File: sdk/package.json
Comment (yugure-orca) [inline sdk/package.json:1]: On my investigation, ``@metaplex-foundation/mpl-token-metadata`` is NOT used in ``src``.  it is used in ``tests`` so we can move this package to devDependencies.

Also, this old package depends on spl-token v0.1.8.
It is causing multiple versions of spl-token to get mixed up.
v0.1.8 can be completely eliminated by upgrading mpl-token-metadata to 2.12.0 and changing Metadata.load to Metadata.Metadata.fromAccountAddress.

File: sdk/src/instructions/composites/swap-with-route.ts [hunk 3]
```diff
@@ -333,13 +337,10 @@ async function cachedResolveOrCreateNonNativeATAs(
...
     } else {
       const createAtaInstruction = createAssociatedTokenAccountInstruction(
-        ASSOCIATED_TOKEN_PROGRAM_ID,
-        TOKEN_PROGRAM_ID,
-        tokenMintArray[index],
+        payer,
         ataAddress,
         ownerAddress,
-        payer,
-        modeIdempotent
```
[the remark this answers is not in the document]
Comment (odcheung) [reply sdk/src/instructions/composites/swap-with-route.ts:342]: ah. we actually shouldn't call the spl-token's instruction directly. I will create a TokenUtil.createAssociatedTokenAccountInstruction to handle the `modeIdempotent` param.

File: sdk/tests/integration/close_position.test.ts [hunk 4]
```diff
@@ -67,7 +68,7 @@ describe("close_position", () => {
       ctx.provider,
       params.positionTokenAccount,
       delegate.publicKey,
-      "CloseAccount",
+      3,
```
[the remark this answers is not in the document]
Comment (odcheung) [reply sdk/tests/integration/close_position.test.ts:71]: good find on that enum!

File: sdk/package.json
Comment (odcheung) [reply sdk/package.json:1]: awesome. with that, now we can completely remove spl-token 1.8!

**Review by yugure-orca — APPROVED** [review]
Thanks for the update!
Now LGTM.

**Pull request merged**